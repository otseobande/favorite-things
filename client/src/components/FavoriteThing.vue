<template>
  <div>
    <div
      :class="`
        w-full
        mt-2
        bg-white
        rounded
        text-gray-700
        p-3
        cursor-pointer
        hover:shadow-2xl
        hover:bg-gray-200
        flex
        flex-row
        justify-between
      `"
      @click="openDetailsModal"
    >
      <span class="m-0 inline capitalize font-bold">
        {{index + 1}}. {{favoriteThing.title}}
      </span>
      <button @click.stop.prevent="confirmDelete" class="p-1 text-center">
        <img src="../assets/cancel.svg" width="10" />
      </button>
    </div>
    <t-modal
      ref="detailsModal"
      closeIconClass="text-gray-300 w-8 right-0 mt-2 mr-6 absolute"
      @closed="editting = false"
    >
      <div v-if="editting">
        <NewFavoriteThing
          :category="category"
          :favoriteThing="favoriteThing"
          :singularizedCategoryName="singularizedCategoryName"
          :editting='editting'
          @finished="editting = false"
        />
      </div>
      <div v-else class="text-gray-700 flex flex-col">
        <div class="text-2xl capitalize font-bold flex items-center" @click="editting = true">
          <span class="">{{favoriteThing.title}}</span>
          <button class="ml-3 border rounded text-sm px-1 text-gray-700 hover:bg-gray-200 flex items-center">
            <span>Edit</span>
            <Pencil class="h-3 w-3 ml-1 inline fill-current"/>
          </button>
        </div>

        <small class="text-gray-600">in category {{category.name}}</small>
        <div class="mt-4">
          <h3 class="text-lg font-bold">Description</h3>
          <div>
            <div v-if="favoriteThing.description" class="description">
              {{favoriteThing.description}}
            </div>
            <div v-else class="text-sm italic text-gray-500">
              Not provided
            </div>
          </div>
        </div>
        <div class="mt-3">
          <h3 class="text-lg font-bold">Ranking</h3>
          <p>
            {{rankingWithOrdinal}}
          </p>
        </div>
        <div v-if="metadata">
          <div class="mt-3" v-for="(data, index) in metadata" :key="index">
            <h3 class="text-lg font-bold">{{data.label}}</h3>
            <p v-if="Array.isArray(data.value)">
              <ul class="list-disc pl-8">
                <li v-for="(value, index) in data.value" :key="index">
                  {{value}}
                </li>
              </ul>
            </p>
            <p v-else>
              {{data.value}}
            </p>
          </div>
        </div>
        <div class="flex flex-col mt-3 text-sm">
          <t-button size="sm" @click="viewActivities = !viewActivities">
            <span v-if="viewActivities">Close activities</span>
            <span v-else>View activities</span>
          </t-button>
          <template v-if="viewActivities">
            <div
              class="mt-2 border-b"
              v-for="event in updateHistory"
              :key="event.timestamp"
            >
              <p>{{getChangesString(event.changes)}}</p>
              <p class="italic text-gray-600 mt-1">on {{formatDate(event.timestamp)}}</p>
            </div>
          </template>
        </div>
      </div>

    </t-modal>
    <t-modal ref="deleteModal" closeIconClass="text-gray-300 w-8 right-0 mt-2 mr-6 absolute">
      <div class="text-gray-700">
        <h2 class="text-lg mt-16 text-center">
          Add you sure you want to delete {{favoriteThing.title}}?
        </h2>
        <div class="mt-6 flex justify-center">
          <div>
            <t-button
              variant="danger"
              size="sm"
              class="mr-2 font-bold"
              type='submit'
              @click="deleteFavoriteThing"
            >
              Delete
            </t-button>
            <t-button size="sm" class="font-bold" @click="closeConfirmDeleteModal">
              Cancel
            </t-button>
          </div>
        </div>
      </div>
    </t-modal>
  </div>
</template>

<script>
// Vuex
import { mapActions } from 'vuex';
import * as actionTypes from '../store/actions.type';

import toast from 'toast-me';

import getNumberWithOrdinal from '../utils/getNumberWithOrdinal';
import format from 'date-fns/format';

// components
import Pencil from './Pencil.vue';
import NewFavoriteThing from './NewFavoriteThing.vue'

export default {
  props: {
    favoriteThing: {
      required: true,
      type: Object
    },
    index: {
      required: true,
      type: Number
    },
    category: {
      required: true,
      type: Object
    },
    singularizedCategoryName: {
      required: true,
      type: String
    }
  },
  data() {
    return {
      showConfirmDeleteModal: false,
      editting: false,
      viewActivities: false
    }
  },
  components: {
    Pencil,
    NewFavoriteThing
  },
  methods: {
    ...mapActions([
      actionTypes.DELETE_FAVORITE_THING
    ]),
    confirmDelete() {
      this.$refs.deleteModal.show()
    },
    closeConfirmDeleteModal() {
      this.$refs.deleteModal.hide()
    },
    openDetailsModal() {
      this.$refs.detailsModal.show()
    },
    closeDetailsModal() {
      this.$refs.detailsModal.hide();
      this.editting = false;
    },
    async deleteFavoriteThing(){
      this.closeConfirmDeleteModal();
      await this[actionTypes.DELETE_FAVORITE_THING]({
        favoriteThingId: this.favoriteThing.id
      });

      toast('Favorite thing deleted successfully!');
    },
    getChangesString(changes) {
      const parsedChanges = JSON.parse(changes);

      const changesString = Object.keys(parsedChanges).reduce((acc, key, index, arr) => {
        const change = parsedChanges[key];
        let seperator = '';

        if (index < arr.length - 2) {
          seperator = ', ';
        }

        if (index === arr.length - 2) {
          seperator = ' and ';
        }

        if (index === arr.length - 1) {
          seperator = '.'
        }

        return `${acc}${key} from ${change[0]} to ${change[1]}${seperator}`
      }, 'Changed ');

      return changesString;
    },
    formatDate(date) {
      return format(new Date(date), 'Do MMM, YYYY. h:mma')
    }
  },
  computed: {
    rankingWithOrdinal () {
      return getNumberWithOrdinal(this.favoriteThing.ranking)
    },
    metadata () {
      if (typeof(this.favoriteThing.metadata) === 'object') {
        return this.favoriteThing.metadata;
      }

      return JSON.parse(this.favoriteThing.metadata);
    },
    updateHistory () {
      return this.favoriteThing.history.filter(event => {
        return event.action === 1;
      })
    }
  }
}
</script>

<style>
.description::first-letter {
  text-transform: uppercase;
}
</style>
