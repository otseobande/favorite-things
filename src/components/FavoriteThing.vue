<template>
  <div class='w-full ml-2 mt-2 bg-white rounded shadow-md text-gray-700 p-3 cursor-pointer hover:shadow-xl'>
    <h3 class="m-0 mb-3 inline">{{favoriteThing.name}}</h3>
  </div>
</template>

<script>
export default {
  props: ['favoriteThing']
}
</script>
