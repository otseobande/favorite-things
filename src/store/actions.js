import * as actionTypes from './actions.type';
import * as mutationTypes from './mutations.type';

import categoryService from '../services/categoryService';
import favoriteThingService from '../services/favoriteThingService';

export default {
  async [actionTypes.FETCH_CATEGORIES]({ commit }) {
    const response = await categoryService.get();

    commit(mutationTypes.SET_CATEGORIES, response.data.results)
  },

  async [actionTypes.ADD_FAVORITE_THING]({ commit }, newFavoriteThing) {
    const response = await favoriteThingService.create(newFavoriteThing);

    commit(mutationTypes.ADD_FAVORITE_THING, response.data)
  },

  async [actionTypes.UPDATE_FAVORITE_THING]({ commit }, {
    favoriteThingId,
    updatedFavoriteThingData
  }) {
    const response = await favoriteThingService.update({
      favoriteThingId,
      updatedFavoriteThingData
    })

    commit(mutationTypes.UPDATE_FAVORITE_THING, response.data);
  },

  async [actionTypes.DELETE_FAVORITE_THING]({ commit }, { favoriteThingId }) {
    await favoriteThingService.delete({
      favoriteThingId
    });

    commit(mutationTypes.DELETE_FAVORITE_THING, favoriteThingId)
  },

  async [actionTypes.ADD_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.ADD_CATEGORY, {
      id: 'template',
      name: '',
      favorite_things: []
    })
  },

  async [actionTypes.DELETE_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.DELETE_CATEGORY, 'template')
  },

  async [actionTypes.ADD_CATEGORY]({ commit }, newCategory) {
    const response = await categoryService.create(newCategory)

    commit(mutationTypes.ADD_CATEGORY, response.data)
  },

  async [actionTypes.DELETE_CATEGORY]({ commit }, { categoryId }) {
    await categoryService.delete(categoryId);

    commit(mutationTypes.DELETE_CATEGORY, categoryId);
  }
}
