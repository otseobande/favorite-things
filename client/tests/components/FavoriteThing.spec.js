import { shallowMount, createLocalVue } from '@vue/test-utils';
import FavoriteThing from '../../src/components/FavoriteThing.vue';
import VueTailwind from 'vue-tailwind';
import Vuex from 'vuex';
import actions from '../../src/store/actions';
import * as actionTypes from '../../src/store/actions.type';
import toast from 'toast-me';
import format from 'date-fns/format';

jest.mock('../../src/store/actions');
jest.mock('date-fns/format');

const localVue = createLocalVue();

localVue.use(VueTailwind);
localVue.use(Vuex);

let wrapper;
const propsData = {
  favoriteThing: {
    title: 'Test thing',
    ranking: 1,
    category: 1,
    metadata: [],
    history: [
      {
        action: 1,
        changes: {
          'name': ['People', 'Places'],
          'title': ['food', 'sport'],
          'job': ['cleaner', 'software developer']
         }
      },
      {
        action: 2,
        changes: {
          'name': ['People', 'Places'],
          'title': ['food', 'sport'],
          'job': ['cleaner', 'software developer']
         }
      }
    ]
  },
  index: 0,
  category: {
    name: 'Test category'
  },
  singularizedCategoryName: 'Test category'
}

describe('FavoriteThing component', () => {
  beforeEach(() => {
    wrapper = shallowMount(FavoriteThing, {
      localVue,
      store: new Vuex.Store({
        actions
      }),
      propsData,
    });
  })
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('confirmDelete method', () => {
    it('should show delete modal on call', () => {
      wrapper.vm.$refs.deleteModal = {
        show: jest.fn()
      };

      wrapper.vm.confirmDelete();

      expect(wrapper.vm.$refs.deleteModal.show).toBeCalled();
    })
  });

  describe('closeConfirmDeleteModal', () => {
    it('should close confirm delete modal', () => {
      wrapper.vm.$refs.deleteModal = {
        hide: jest.fn()
      };

      wrapper.vm.closeConfirmDeleteModal();

      expect(wrapper.vm.$refs.deleteModal.hide).toBeCalled();
    })
  });

  describe('openDetailsModal', () => {
    it('should open the details modal', () => {
      wrapper.vm.$refs.detailsModal = {
        show: jest.fn()
      };

      wrapper.vm.openDetailsModal();

      expect(wrapper.vm.$refs.detailsModal.show).toBeCalled();
    })
  });

  describe('closeDetailsModal', () => {
    it('should close the details modal', () => {
      wrapper.vm.$refs.detailsModal = {
        hide: jest.fn()
      };

      wrapper.vm.closeDetailsModal();

      expect(wrapper.vm.$refs.detailsModal.hide).toBeCalled();
      expect(wrapper.vm.editting).toEqual(false);
    });
  });

  describe('deleteFavoriteThing', () => {
    it('should close confirmDeleteModal and call action method', async () => {
      wrapper.vm.closeConfirmDeleteModal = jest.fn();
      wrapper.vm[actionTypes.DELETE_FAVORITE_THING] = jest.fn();

      await wrapper.vm.deleteFavoriteThing();
      expect(wrapper.vm[actionTypes.DELETE_FAVORITE_THING]).toBeCalledWith(
        {
          favoriteThingId: wrapper.vm.favoriteThing.id
        }
      )
      expect(toast).toBeCalledWith('Favorite thing deleted successfully!');
      toast.mockReset()
    });
  });

  describe('getChangesString', () => {
    it('should parse changes json string to a human readable string', async () => {
      const changes = {
        'name': ['People', 'Places'],
        'title': ['food', 'sport'],
        'job': ['cleaner', 'software developer']
       };

      const changesString = wrapper.vm.getChangesString(JSON.stringify(changes));

      expect(changesString).toEqual('Changed name from People to Places, title from food to sport and job from cleaner to software developer.')
    });
  });

  describe('formatDate method', () => {
    it('should use datefns to format date', () => {
      wrapper.vm.formatDate('2018-09-09');

      expect(format).toBeCalledWith(new Date('2018-09-09'), 'Do MMM, YYYY. h:mma')
    })
  });

  describe('updateHistory computed field', () => {
    it('should return events with action 1', () => {
      expect(wrapper.vm.updateHistory.every(event => {
        return event.action == 1;
      })).toBeTruthy();
    })
  })

  describe('metadata computed field', () => {
    it('should return parsed metadata', () => {
      wrapper.setProps({
        favoriteThing: {
          ...propsData.favoriteThing,
          metadata: '[]'
        }
      })
      expect(wrapper.vm.metadata).toEqual([]);
    })
  })
});
