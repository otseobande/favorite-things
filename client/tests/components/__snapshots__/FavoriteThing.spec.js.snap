// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteThing component should match snapshot 1`] = `
<div>
  <div class="
      w-full
      mt-2
      bg-white
      rounded
      text-gray-700
      p-3
      cursor-pointer
      hover:shadow-2xl
      hover:bg-gray-200
      flex
      flex-row
      justify-between
    "><span class="m-0 inline capitalize font-bold">
      1. Test thing
    </span> <button class="p-1 text-center"><img src="../assets/cancel.svg" width="10"></button></div>
  <t-modal-stub clicktoclose="true" esctoclose="true" width="600" height="300" pivoty="0.1" baseclass="z-40 overflow-auto left-0 top-0 bottom-0 right-0 w-full h-full fixed" overlayclass="z-40 overflow-auto left-0 top-0 bottom-0 right-0 w-full h-full fixed bg-black opacity-75" containerclass="z-50 relative p-3 mx-auto my-0 max-w-full" wrapperclass="bg-white rounded shadow-lg border flex flex-col overflow-hidden" bodyclass="p-6 flex-grow" headerclass="px-6 py-3 text-xl border-b" footerclass="px-6 py-3 border-t" closeiconclass="text-gray-300 w-8 right-0 mt-2 mr-6 absolute">
    <div class="text-gray-700 flex flex-col">
      <div class="text-2xl capitalize font-bold flex items-center"><span>Test thing</span> <button class="ml-3 border rounded text-sm px-1 text-gray-700 hover:bg-gray-200 flex items-center"><span>Edit</span>
          <pencil-stub class="h-3 w-3 ml-1 inline fill-current"></pencil-stub></button></div> <small class="text-gray-600">in category Test category</small>
      <div class="mt-4">
        <h3 class="text-lg font-bold">Description</h3>
        <div>
          <div class="text-sm italic text-gray-500">
            Not provided
          </div>
        </div>
      </div>
      <div class="mt-3">
        <h3 class="text-lg font-bold">Ranking</h3>
        <p>
          1st
        </p>
      </div>
      <div></div>
      <div class="flex flex-col mt-3 text-sm">
        <t-button-stub tagname="button" type="button" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm"><span>View activities</span></t-button-stub>
        <!---->
      </div>
    </div>
  </t-modal-stub>
  <t-modal-stub clicktoclose="true" esctoclose="true" width="600" height="300" pivoty="0.1" baseclass="z-40 overflow-auto left-0 top-0 bottom-0 right-0 w-full h-full fixed" overlayclass="z-40 overflow-auto left-0 top-0 bottom-0 right-0 w-full h-full fixed bg-black opacity-75" containerclass="z-50 relative p-3 mx-auto my-0 max-w-full" wrapperclass="bg-white rounded shadow-lg border flex flex-col overflow-hidden" bodyclass="p-6 flex-grow" headerclass="px-6 py-3 text-xl border-b" footerclass="px-6 py-3 border-t" closeiconclass="text-gray-300 w-8 right-0 mt-2 mr-6 absolute">
    <div class="text-gray-700">
      <h2 class="text-lg mt-16 text-center">
        Add you sure you want to delete Test thing?
      </h2>
      <div class="mt-6 flex justify-center">
        <div>
          <t-button-stub tagname="button" type="submit" variant="danger" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm" class="mr-2 font-bold"><span>Delete</span></t-button-stub>
          <t-button-stub tagname="button" type="button" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm" class="font-bold">
            Cancel
          </t-button-stub>
        </div>
      </div>
    </div>
  </t-modal-stub>
</div>
`;
